app/mqtt_client.py

```diff
@@ -24,11 +26,15 @@ def run(self) -> None:
         """
         this is a blocking operation.
         """
-        self.client.on_connect = self._on_connect()
-        self.client.on_message = self._on_message()
+        self._client.on_connect = self._on_connect()
+        self._client.on_message = self._on_message()
+
+        try:
+            self._client.connect(self.host, self.port)
+        except OSError as e:
+            raise OSError(f"Cannot assign requested address: {self.host}:{self.port}") from e
```

Comment:
this error message is missing and is actually quite annoying its absence. It was the motivation for my change, but I ended up adding more test improvements as well.